src/main/java/com/hivemq/persistence/local/memory/ClientSessionSubscriptionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,216 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static com.hivemq.util.ThreadPreConditions.SINGLE_WRITER_THREAD_PREFIX;
+
+/**
+ * A memory based ClientSessionSubscriptionLocalPersistence.
+ *
+ * @author Florian Limpöck
+ */
+@LazySingleton
+public class ClientSessionSubscriptionMemoryLocalPersistence implements ClientSessionSubscriptionLocalPersistence {
+
+    private final @NotNull Map<String, IterablePersistenceEntry<ImmutableSet<Topic>>> @NotNull [] buckets;
+    private final int bucketCount;
+
+    @VisibleForTesting
+    final @NotNull AtomicLong currentMemorySize = new AtomicLong();
+
+    @Inject
+    ClientSessionSubscriptionMemoryLocalPersistence(@NotNull final MetricRegistry metricRegistry) {
+
+        bucketCount = InternalConfigurations.PERSISTENCE_BUCKET_COUNT.get();
+
+        //noinspection unchecked
+        buckets = new HashMap[bucketCount];
+        for (int i = 0; i < bucketCount; i++) {
+            buckets[i] = new HashMap<>();
+        }
+
+        metricRegistry.register(HiveMQMetrics.CLIENT_SESSION_SUBSCRIPTIONS_MEMORY_PERSISTENCE_TOTAL_SIZE.name(),
+                (Gauge<Long>) currentMemorySize::get);
+
+    }
+
+    @Override
+    @ExecuteInSingleWriter
+    public void addSubscription(
+            @NotNull final String client,
+            @NotNull final Topic topic,
+            final long timestamp,
+            final int bucketIndex) {
+        addSubscriptions(client, ImmutableSet.of(topic), timestamp, bucketIndex);
+    }
+
+    @Override
+    @ExecuteInSingleWriter
+    public void addSubscriptions(
+            @NotNull final String client,
+            @NotNull final ImmutableSet<Topic> topics,
+            final long timestamp,
+            final int bucketIndex) {
+        checkNotNull(client, "Client id must not be null");
+        checkNotNull(topics, "Topics must not be null");
+        checkState(timestamp > 0, "Timestamp must not be 0");
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+        buckets[bucketIndex].compute(client, (ignore, oldEntry) -> {
+            if (oldEntry == null) {
+                final IterablePersistenceEntry<ImmutableSet<Topic>> newEntry = new IterablePersistenceEntry<>(topics, timestamp);
+                currentMemorySize.addAndGet(newEntry.getEstimatedSize());
+                currentMemorySize.addAndGet(ObjectMemoryEstimation.stringSize(client));
+                return newEntry;
+            }
+            currentMemorySize.addAndGet(-oldEntry.getEstimatedSize());
+            final IterablePersistenceEntry<ImmutableSet<Topic>> mergedEntry =
+                    new IterablePersistenceEntry<>(Sets.union(topics, oldEntry.getObject()).immutableCopy(), timestamp);
+            currentMemorySize.addAndGet(mergedEntry.getEstimatedSize());
```

Comment:
minor improvement: you could also run the size estimation by summing along the new topic set and initializing the entry with the old entry's size + new topics size; not sure if the impact is that great usually though...

Reply:
but i dont have the intersection of both sets.